contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -1070,15 +1070,69 @@ validate_special_function(char *func_nsname, char *func_name, int nargs, Oid *in
 
 	/* If function is not a special function no additional handling required */
 	if (special_func == NULL)
+	{
+		/* report error for case when NULL casted to different datatypes and passed as 2nd or 3rd argument of SUBSTRING() function */
+		if (strlen(func_name) == 9 && strncmp(func_name, "substring", 9) == 0)
+		{
+			for (int i = 1; i < nargs; i++)
+			{
+				Node *arg = (Node *) lfirst(list_nth_cell(fargs, i));
+
+				if (input_typeids[i] == UNKNOWNOID)
+      				continue;
+
+				/* Throw error when input is constant and NULL */
+				if (IsA(arg, Const) && ((Const *)arg)->constisnull)
+				{
+					char	*typ_name = NULL;
+					int		len;
+
+					if (common_utility_plugin_ptr == NULL)
+						ereport(ERROR,
+								(errcode(ERRCODE_INTERNAL_ERROR),
+									errmsg("Failed to find common utility plugin.")));
+
+					(*common_utility_plugin_ptr->resolve_pg_type_to_tsql) (input_typeids[i], &typ_name);
+					if(typ_name)
+					{
+						len = strlen(typ_name);
+
+						if (!((len == 3 && strncmp(typ_name,"int", 3) == 0) ||
+							(len == 7 && strncmp(typ_name,"tinyint", 7) == 0) ||
+							(len == 8 && strncmp(typ_name,"smallint", 8) == 0) ||
+							(len == 6 && strncmp(typ_name,"bigint", 6) == 0)))
+							ereport(ERROR,
+								(errcode(ERRCODE_UNDEFINED_FUNCTION),
+									errmsg("Argument data type %s is invalid for argument %d of substring function.", 
+											format_type_be(input_typeids[i]), i+1)));
+
+						pfree(typ_name);
+					}
+				}
+			}
+		}
 		return false;
+	}		
 
 	sys_varcharoid = get_sys_varcharoid();
 
 	/* Report error in case of invalid argument datatype */
 	for (int i = 0; i < special_func->nargs; i++)
 	{
-		/* for unknown literals consider its type as sys.VARCHAR */
-		input_type_id = (input_typeids[i] == UNKNOWNOID) ? sys_varcharoid : input_typeids[i];
+		/* 
+		 * if argument is NULL then keep its typeId as UNKNOWN and skip the report error handling 
+		 * otherwise consider it as sys.VARCHAR
+		 */
```

Comment:
which exact test is affecting due to this change?

Reply:
Consider the EXAMPLE:
```
Select substring('abc', NULL, 2)
```
in this, 2nd Arguments should be of type int, but according to current workflow, we will consider it as Varchar,
So ideally it should return `NULL` but it will return Error that varchar is not accepted for argument 2, so to handle this we added this change.